### ISSUE TITLE
"Expression" object has no attribute "diff", how to properly use it?

### ISSUE THREAD

**Issue by CodeGodz** [vaexio/vaex#2011]
Hey all, first of all, cool package!

**Code**
```python
import vaex
import numpy as np
import pandas as pd

# Some random data
data = {
    'x': np.random.randint(5, size=20),
    'y': np.random.randint(50, size=20)
}
df = vaex.from_pandas(pd.DataFrame(data))

# Sort the df
# pandas: sorted_df = df.sort_values(by=['x', 'y'])
sorted_df = df.sort(['x', 'y'])

# diff
sorted_df['z'] = (sorted_df['y'].diff() != 1).cumsum().fillna(0)

print(sorted_df.head())
```
Since the `sort` is important for me I added that in the code above, but the simplest way to throw the error is:

**Simpler reproducible example**
```python
df = vaex.from_pandas(pd.DataFrame({'y': np.random.randint(1, size=2)}))
df['z'] = df['y'].diff()
``` 

**Error message**
The code above throws an `AttributeError: 'Expression' object has no attribute 'diff'`

**Software**
```yaml
Python: 3.7.3
Vaex: 3.8.0
```

**Question**
For pandas, when using `sort_values()`, the code works fine but with Vaex I get the `AttributeError`. Probably I'm missing something very obvious...  as based [on this](https://github.com/vaexio/vaex/issues/955) `diff` must be possible.

**Comment by CodeGodz** [comment]
I see it might work with `sorted_df.diff(periods=2, column='y')` but how can I still use that in the case of `(sorted_df['y'].diff() != 1).cumsum().fillna(0)` like in Pandas?

**Comment by maartenbreddels** [comment]
Hi,

Indeed, as mentioned in https://github.com/vaexio/vaex/issues/955#issuecomment-877180508  diff is a dataframe method.

> but how can I still use that in the case of `(sorted_df['y'].diff() != 1).cumsum().fillna(0)` like in Pandas?

sorry, I don't know what the question is, can you elaborate?

**Comment by CodeGodz** [comment]
Hey @maartenbreddels! We have a dataframe of around 8 billion rows and do a join with around 10K. Currently comparing how different tools perform, pandas, dask, polars, pyarrow, etc. Currently we were not able to perform the operations we wanted: in-memory ones ran out of memory (also for a 100M join), dask was quite slow and pyarrow took ages to just process the input, so now I'm trying Vaex. The file load and join work super fast with `Vaex` - and that's a big join :). However, now I'm stuck at a piece of code I could write in pandas and polars but not in Vaex.

**Code goal**
This piece of code basically assigns an unique group id, `z`, based on a sort + adjacent `y`  values (hence the `diff()`) if the difference `!=1` they are not adjacent (anymore) hence we should increase our id, doing this using `cumsum()` as the `True` will be cumulatively added creating an auto incrementing id `z` every time the difference is `!=1`  (See example below).

**Example of generating the ID with Pandas**
Let me demonstrate with a clearer example, lets take this dataframe:
```
data = {
    'x': [1,2,3,4, 5, 6,  7,8,9,10],
    'y': [1,2,3,10,11,12,30,31, 0, 100]
}
df = pd.DataFrame(data)
``` 
If we then run:
`df['z'] = (df['y'].diff() != 1).cumsum().fillna(0)` 

We will get an unique id per "group" of adjacent y values:
```
    x    y  z
0   1    1  1
1   2    2  1
2   3    3  1
3   4   10  2
4   5   11  2
5   6   12  2
6   7   30  3
7   8   31  3
8   9    0  4
9  10  100  5
``` 

**The question :)**
In short, how can I do the same in Vaex?

The longer version, I can do `df = df.diff(column = 'y')` like:
```
df = df.diff(column = 'y')
df['z'] = (df['y'] != 1)
``` 
First this would replace the `y` column, I want to keep it, and then I get stuck as Vaex does not have the `cumsum()`, I could perhaps use `np.cumsum()` but not sure if that will break Vaex - which is pretty relevant for 8 billion rows haha. 

Hope it's a bit more clear now, if not let me know! :)

**Comment by maartenbreddels** [comment]
You can do:
```python
df['y_shift'] = df['y']
df = df.diff(column='y_shift'])
```

We don't have a lazy cumsum yet (funding would be welcome for that).

In the meantime, you can do
```python
df['z'] = (df['y_shift'] != 1).to_numpy().cumsum()
```
This would materialize it, so it would use some RAM, but since you have plenty of that :)


Does this help?

**Comment by CodeGodz** [comment]
Sorry missed this notification, yess that helps :) Thanks Maarten!
